### ISSUE TITLE
What is the best way for the inference process in LORA in PEFT approach

### ISSUE THREAD

**Issue by pradeepdev-1995** [huggingface/peft#1310]
### Feature request

What is the best way for the inference process in LORA in PEFT approach
### Motivation

What is the best way for the inference process in LORA in PEFT approach
### Your contribution

Here is the SFTtrainer method i used for finetuning mistral
```
trainer = SFTTrainer(
    model=peft_model,
    train_dataset=data,
    peft_config=peft_config,
    dataset_text_field=" column name",
    max_seq_length=3000,
    tokenizer=tokenizer,
    args=training_arguments,
    packing=packing,
)
trainer.train()
```
I found different mechanisms for the finetuned model inference after PEFT based LORA finetuning

Method - 1

save adapter after completing training and then merge with base model then use for inference
```
trainer.model.save_pretrained("new_adapter_path")
from peft import PeftModel
finetuned_model = PeftModel.from_pretrained(base_model,
                                  new_adapter_path,
                                  torch_dtype=torch.float16,
                                  is_trainable=False,
                                  device_map="auto"
                                  )
finetuned_model = finetuned_model.merge_and_unload()
``` 

Method - 2

save checkpoints during training and then use the checkpoint with the least loss
```
from peft import PeftModel
finetuned_model = PeftModel.from_pretrained(base_model,
                                  "least loss checkpoint path",
                                  torch_dtype=torch.float16,
                                  is_trainable=False,
                                  device_map="auto"
                                  )
finetuned_model = finetuned_model.merge_and_unload()
``` 
Method - 3

same method with AutoPeftModelForCausalLM class 
```
model = AutoPeftModelForCausalLM.from_pretrained(
    "output directory checkpoint path",
    low_cpu_mem_usage=True,
    return_dict=True,
    torch_dtype=torch.float16,
    device_map="cuda")
finetuned_model = finetuned_model.merge_and_unload()
```
Method-4

AutoPeftModelForCausalLM class specifies the output folder without specifying a specific checkpoint
```
instruction_tuned_model = AutoPeftModelForCausalLM.from_pretrained(
    training_args.output_dir,
    torch_dtype=torch.bfloat16,
    device_map = 'auto',
    trust_remote_code=True,
)
finetuned_model = finetuned_model.merge_and_unload()
```
Method-5
All the above methods without merging
```
#finetuned_model = finetuned_model.merge_and_unload()
```

Which is the actual method I should follow for inference?
and when to use which method over another?

**Comment by BenjaminBossan** [comment]
Regarding 1 vs 2: Whether to load the last or the best checkpoint depends on your use case and is not PEFT-specific. Generally, the best checkpoint is expected to work better than the last checkpoint if your model starts overfitting, but choosing that one could lead to stronger "overfitting on the validation dataset" (if you use the validation dataset to determine the "best" checkpoint).

Regardign 1 vs 3: It doesn't really matter here if you use `PeftModel` or `AutoPeftModelForCausalLM`, if you check the type of the model after loading, it should be `PeftModelForCausalLM` in both cases.

Regarding 4: I don't understand this case.

Regarding 5: There are advantages and disadvantages to using `merge_and_unload`. In a nutshell, the advantage should be that it runs inference faster. Cons: You cannot unmerge or disable the adapter if you want to run inference on the base model, you cannot add more adapters, merging can introduce very small error, etc. If you only want to run inference with the one adapter, `merge_and_unload` should be the choice.

**Comment by pradeepdev-1995** [comment]
@BenjaminBossan Thank you very much. 
Regarding 4: here we are specifying the output directory without specifying any particular checkpoint folder inside it. i found this method somewhere

So which is the actual method we should follow for inference?

**Comment by BenjaminBossan** [comment]
> Regarding 4: here we are specifying the output directory without specifying any particular checkpoint folder inside it.

Hmm, I can't comment on that, since I don't know what exactly the `Trainer` does, but I'd rather be explicit.

> So which is the actual method we should follow for inference?

I can't decide this for you, as it depends on the use case. My last reply contained some suggestions for you to take into account when making the decision.

**Comment by pradeepdev-1995** [comment]
Thank you @BenjaminBossan

**Comment by BenjaminBossan** [comment]
I hope this helps you decide :) I'll close the issue for now, feel free to re-open if you have more questions.
